File: algorithms/algorithms-python/leetcode/Question_0037_Sudoku_Solver.py
#!/usr/bin/python
# coding: utf-8

class Solution(object):
    def solveSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: void Do not return anything, modify board in-place instead.
        """
        if not board or len(board) != 9 or len(board[0]) != 9:
            return
        self.help(board, 0, 0)

    def help(self, board, i, j):
        if i >= 9:
            return True
        if j >= 9:
            return self.help(board, i + 1, 0)
        if board[i][j] == '.':
            for m in '123456789':
                board[i][j] = m
                if self.isValid(board, i, j) and self.help(board, i, j + 1):
                    return True
                board[i][j] = '.'
        else:
            return self.help(board, i, j + 1)
        return False


    def isValid(self, board, i, j):
        for m in range(9):
            if m != i and board[i][j] == board[m][j]:
                return False
            if m != j and board[i][j] == board[i][m]:
                return False
        row, col = i / 3 * 3, j / 3 * 3
        for m in range(row, row + 3):
            for n in range(col, col + 3):
                if (m != i or n != j) and board[i][j] == board[m][n]:
                    return False
        return True

File: algorithms/algorithms-cplusplus/leetcode/Question_0037_Sudoku_Solver.cc
#include <vector>

using namespace std;

class Solution {
 public:
  void solveSudoku(vector<vector<char>>& board) {
    if (board.size() != 9 || board[0].size() != 9) {
      return;
    }
    help(board, 0, 0);
  }

  bool help(vector<vector<char>>& board, int r, int c) {
    if (r >= 9) {
      return true;
    }
    if (c >= 9) {
      return help(board, r + 1, 0);
    }
    if (board[r][c] != '.') {
      return help(board, r, c + 1);
    } else {
      for (int n = 1; n <= 9; n++) {
        board[r][c] = n + '0';
        if (isValid(board, r, c) && help(board, r, c + 1)) {
          return true;
        }
      }
      board[r][c] = '.';
    }
    return false;
  }

  bool isValid(vector<vector<char>>& board, int r, int c) {
    for (int i = 0; i < 9; i++) {
      if (i != r && board[r][c] == board[i][c]) {
        return false;
      }
    }
    for (int j = 0; j < 9; j++) {
      if (j != c && board[r][c] == board[r][j]) {
        return false;
      }
    }
    int m = r / 3 * 3, n = c / 3 * 3;
    for (int i = m; i < m + 3; i++) {
      for (int j = n; j < n + 3; j++) {
        if ((i != r || j != c) && board[i][j] == board[r][c]) {
          return false;
        }
      }
    }
    return true;
  }
};
